getting_started/first_3d_game/04.mob_scene.rst

```diff
@@ -0,0 +1,231 @@
+player using the "look_at() "method and randomize the angle by rotating a
+random amount around the Y axis. Below, ``rand_range()`` outputs a random value
+between ``-PI / 4`` radians and ``PI / 4`` radians.
+
+::
+
+   # We will call this function from the Main scene.
+   func initialize(start_position, player_position):
+       translation = start_position
+       # We turn the mob so it looks at the player.
+       look_at(player_position, Vector3.UP)
+       # And rotate it randomly so it doesn't move exactly toward the player.
+       rotate_y(rand_range(-PI / 4, PI / 4))
+
+We then calculate a random speed using "rand_range() "once again and we use it
+to calculate the velocity.
+
+We start by creating a 3D vector pointing forward, multiply it by our
+"random_speed ", and finally rotate it using the "Vector3 "class's
+``rotated()`` method.
+
+::
+
+       # We calculate a random speed.
+       var random_speed = rand_range(min_speed, max_speed)
+       # We calculate a forward velocity that represents the speed.
+       velocity = Vector3.FORWARD * random_speed
+       # We then rotate the vector based on the mob's Y rotation to move in the direction it's looking.
+       velocity = velocity.rotated(Vector3.UP, rotation.y)
+
+Leaving the screen
+------------------
+
+We still have to destroy the mobs when they leave the screen. To do so, we'll
+connect our *VisibilityNotifier* node's ``screen_exited`` signal to the *Mob*.
+
+Head back to the 3D viewport by clicking on the *3D* label at the top of the
+editor. You can also press F2.
+
+|image8|
+
+Select the *VisibilityNotifier* node and on the right side of the interface,
+navigate to the *Node* dock. Double-click the *screen_exited()* signal.
+
+|image9|
+
+Connect the signal to the *Mob*.
+
+|image10|
+
+This will take you back to the script editor and add a new function for you,
+``_on_VisibilityNotifier_screen_exited()``. From it, call the ``queue_free()``
+method. This will destroy the mob instance when the *VisibilityNotifier* \'s box
+leaves the screen.
+
+::
+
+   func _on_VisibilityNotifier_screen_exited():
+       queue_free()
+
+Our monster is ready to enter the game! In the next part, you will spawn
+monsters in the game level.
+
+Here is the complete "Mob.gd "script for reference.
+
+::
+
+   extends KinematicBody
+
+   # Minimum speed of the mob in meters per second.
+   export var min_speed = 10
+   # Maximum speed of the mob in meters per second.
+   export var max_speed = 18
+
+   var velocity = Vector3.ZERO
+
+
+   func _physics_process(_delta):
+       move_and_slide(velocity)
+
+   func initialize(start_position, player_position):
+       translation = start_position
+       look_at(player_position, Vector3.UP)
+       rotate_y(rand_range(-PI / 4, PI / 4))
+
+       var random_speed = rand_range(min_speed, max_speed)
+       velocity = Vector3.FORWARD * random_speed
+       velocity = velocity.rotated(Vector3.UP, rotation.y)
```

Comment:
```suggestion
       # Give the mob forward movement. +Z is back, so -Z is forward.
       velocity = -transform.basis.z * random_speed
```

Reply:
Teaching-wise, basis adds complexity because it encodes both rotation and scale in 3 vectors. While rotating the forward vector to match the player's rotation only takes rotation into account, and it's explicit.

This change would require explaining transform + basis for the user to understand what's happening here. As this is supposed to be a tutorial for people who've never done any 3D before, I've chosen not to use it here, even if it's a common and useful tool. I would keep it for examples in the 3D section, that goes more in-depth.

Reply:
Yes, but I think the comment explains it quite well, so I'm still in favor of what I proposed. Your call though.

Another option would be `velocity = transform.basis.xform(Vector3.FORWARD) * random_speed` if you want to avoid the details of which direction is forward in the basis.

Reply:
By "explaining transform + basis", I mean explaining what a basis is and how it works, in general. The comment doesn't do that. An example would be the intro to transforms page in the math section. This would be a good article to read before working with transforms.

You have to consider the target audience. Limited math skills (some remnants of vector math classes in middle or high school), some programming foundations but not necessarily much experience yet, writing their first 3D game. It's not a stage at which you integrate best practices, especially not until you fully understand why you should favor one approach over another.